src/dali_executor/dali_pipeline.cc

```diff
@@ -70,11 +70,20 @@ void DaliPipeline::SetInput(const void* ptr, const char* name, device_type_t sou
            input_shape.sample_dim());
 }
 
+void DaliPipeline::SetInput(const IDescr& io_descr) {
+  ENFORCE(io_descr.buffers.size() == 1, "DALI pipeline input has to be a single chunk of memory");
+  auto meta = io_descr.meta;
+  auto buffer = io_descr.buffers[0];
+  SetInput(buffer.data, meta.name.c_str(), buffer.device, meta.type, meta.shape);
+}
+
+
 void DaliPipeline::PutOutput(void* destination, int output_idx, device_type_t destination_device) {
   assert(destination != nullptr);
   assert(output_idx >= 0);
-  daliOutputCopy(&handle_, destination, output_idx, destination_device,
-                 destination_device == CPU ? nullptr : output_stream_, 0);
+  DeviceGuard dg(device_id_);
+  daliOutputCopy(&handle_, destination, output_idx, destination_device, output_stream_, 0);
+  CUDA_CALL(cudaStreamSynchronize(output_stream_));
```

Comment:
DO we want to sync after each copy? Maybe we can do it once after all copies have been scheduled?

Reply:
Moved synchronization to another function that is called after copying all of the outputs.